// vue-daum-postcode/lib/vue-daum-postcode.mjs
import { defineComponent as r, h as s } from "vue";
function l(a) {
  let t = null;
  return () => {
    try {
      return t || (t = Promise.resolve(a()).catch((o) => {
        throw t = null, o;
      }));
    } catch (o) {
      return Promise.reject(o);
    }
  };
}
var i = (a, t) => new Promise((o, e) => {
  (document.head || document.body).appendChild(Object.assign(document.createElement("script"), t, { src: a, async: !0, onload: o, onerror: () => e(new Error(`Failed to load script: ${a}`)) }));
});
typeof document > "u" && (i = () => Promise.reject(new Error("load is not supported in nodejs")));
function d(a = {}) {
  const t = a.scriptUrl || "//t1.daumcdn.net/mapjsapi/bundle/postcode/prod/postcode.v2.js", o = l(
    () => i(t).then(
      () => new Promise(
        (e) => window.daum.postcode.load(e)
      )
    )
  );
  return r({
    props: {
      q: {
        type: String,
        default: ""
      },
      animation: {
        type: Boolean,
        default: !1
      },
      noAutoMapping: {
        type: Boolean,
        default: !1
      },
      noShorthand: {
        type: Boolean,
        default: !1
      },
      noSubmitMode: {
        type: Boolean,
        default: !1
      },
      pleaseReadGuide: {
        type: Number,
        default: 0
      },
      pleaseReadGuideTimer: {
        type: Number,
        default: 1.5
      },
      maxSuggestItems: {
        type: Number,
        default: 10
      },
      showMoreHName: {
        type: Boolean,
        default: !1
      },
      hideMapBtn: {
        type: Boolean,
        default: !1
      },
      hideEngBtn: {
        type: Boolean,
        default: !1
      },
      alwaysShowEngAddr: {
        type: Boolean,
        default: !1
      },
      zonecodeOnly: {
        type: Boolean,
        default: !1
      },
      theme: {
        type: Object,
        default: () => ({})
      }
    },
    data() {
      return {
        styleHeight: 0,
        isLoaded: !1
      };
    },
    mounted() {
      o().then(() => (this.isLoaded = !0, this.$nextTick())).then(() => {
        new window.daum.Postcode({
          width: "100%",
          height: "100%",
          animation: this.animation,
          autoMapping: !this.noAutoMapping,
          shorthand: !this.noShorthand,
          pleaseReadGuide: this.pleaseReadGuide,
          pleaseReadGuideTimer: this.pleaseReadGuideTimer,
          maxSuggestItems: this.maxSuggestItems,
          showMoreHName: this.showMoreHName,
          hideMapBtn: this.hideMapBtn,
          hideEngBtn: this.hideEngBtn,
          alwaysShowEngAddr: this.alwaysShowEngAddr,
          zonecodeOnly: this.zonecodeOnly,
          theme: this.theme,
          submitMode: !this.noSubmitMode,
          onsearch: (e) => {
            this.$emit("search", e);
          },
          oncomplete: (e) => {
            this.$emit("complete", e);
          },
          onresize: (e) => {
            this.styleHeight = `${e.height}px`, this.$emit("resize", e);
          }
        }).embed(this.$refs.container, {
          q: this.q,
          autoClose: !1
        }), this.$emit("load");
      }).catch((e) => {
        const n = new Error(`Load ${t} failed.`);
        n.cause = e, this.$emit("error", n);
      });
    },
    render() {
      return s(
        "div",
        {
          class: ["vue-daum-postcode"]
        },
        this.isLoaded || !this.$slots.loading ? [
          s("div", {
            class: ["vue-daum-postcode-container"],
            ref: "container",
            style: {
              height: this.styleHeight
            }
          })
        ] : [
          s(
            "div",
            {
              class: ["vue-daum-postcode-loading"]
            },
            [this.$slots.loading()]
          )
        ]
      );
    }
  });
}
const u = d();
function h(a, { name: t, ...o } = {}) {
  a.component(
    t ?? "VueDaumPostcode",
    Object.keys(o).length > 0 ? d(o) : u
  );
}
const m = {
  install: h
};
export {
  u as VueDaumPostcode,
  d as createVueDaumPostcode,
  m as default,
  h as install
};
